### ISSUE TITLE
Hide/disable low health (red cross) during cutscenes

### ISSUE THREAD

**Issue by Polymega** [elishacloud/Silent-Hill-2-Enhancements#19]
**Update:** After play testing it's been determined using a modifed red cross .png file will eventually freeze/crash the game if it's being shown/used on-screen. So the customized image file for this will not be used.

-------------------------

Hi @AeroWidescreen, 

~~I felt the urge to make the low health (red cross) indicator look nicer (and also reduce it's size by ~~25%~~ about 50% so it's not _as_ intrusive):~~

-----------------------

~~**Side note: "Why reduce its size?"**
It's _very_ big in-game and can be quite a distraction. I minimized it so it's still noticeable but not as intrusive. Why? Because certain endings require you to be at low health for most of the game. So would you really want to play the majority of this game with a giant red cross pulsing the whole time? Talk about a visual nuisance... Remember: The graphic is already crimson red and pulsates; it'll still grab your attention just fine at a smaller size... it's just less annoying now.~~

~~**If you wish to not use this file** simply remove it from the following directory:
`\KONAMI\Silent Hill 2\sh2e\pic\etc\RedCrossIcon.png`~~

-----------------------

Anyway, while testing it out I noticed something rather distracting: It continues to flash/pulsate during in-game cutscenes: 

![sh2pc 2018-08-12 20-30-10](https://user-images.githubusercontent.com/27395260/44008200-a8fe09da-9e6e-11e8-9de0-9b78782bc828.png)

Any chance to have this hidden/disabled during in-game cutscenes? It's already naturally hidden when FMVs play.

**Comment by Polymega** [comment]
This may be the only cutscene where it'd be cool/purposeful to keep it on, but I'd leave that up to you if you'd be willing to look into this request any: 
![sh2pc 2018-08-12 13-01-13](https://user-images.githubusercontent.com/27395260/44004377-dcc15b1c-9e2f-11e8-9b03-09121a4f33dc.png)

**Comment by AeroWidescreen** [comment]
Sure, I'll take a look later. Maybe tonight or tomorrow.

**Comment by AeroWidescreen** [comment]
@Polymega 
https://mega.nz/#!pDgFSRrb!bmmzvhi3Gmtpg1-EcigmS0F9ScF-mqmHaXzm350U2-M


@elishacloud 

I'm using `call 004A02B0` to move the current cutscene number to eax. If the value is either `0x24 `or` 0x00`, I use `pop eax` and continue to the health indicator function. If it's not, then I `pop eax` and` ret`.

```
00476280 jmp codecave

push eax
call 004A02B0 // moves cutscene number to eax
cmp eax,24
je newmem+12 // jumps if cutscene 0x24 "After Hospital Roof Attack"; enables health indicator
cmp eax,00
je newmem+12 // jumps if not a cutscene; enables health indicator
pop eax
ret

// jumps goes here
pop eax 
push ecx
mov eax,[00942348]
jmp 00476286 // continues to health indicator function
```

**Comment by AeroWidescreen** [comment]
@Polymega This needs some testing. Just check a few cutscenes, make sure nothing is missing. Apparently the function I'm modifying is also used by the noise filter and fog, so nulling it results in those two things disappearing.

[Download](https://mega.nz/#!YHxn3biI!8H3uRLI3mvt7I9RDpO-L0McDqTeeD-q5iDYiRUcrwMY)

**Comment by Polymega** [comment]
Thanks, Aero!  

I went through the entirety of both campaigns on v1.0 and also initiated all non-mandatory cutscenes as well. Overall it works greats! The noise filter and fog are there and the red cross is gone when cutscenes play. There were a couple of things I noticed: 

--------------------

Whenever there is a chain of different cutscenes playing after one-another the red cross briefly flashes in between the cutscenes. It's fairly distracting. Some examples of this happening are: 
- Laura tricking you in the hospital
- Hospital Otherworld transition after defeating the Flesh Lips
- Jumping down the many holes in the prison/labyrinth
- James running to the door where Angela is inside the labyrinth ("No! Daddy! Please, dont!") 

This brief red cross flicker also happens when you're in the inventory screen to use an item and then it goes back to the game to use said item. Examples include: 
- Using the metal plates in the fireplace garden in Born From a Wish 
- Using the Key of the Persecuted in the labyrinth

^ For the two situations above I'd wonder if disabling the red cross _during_ transitions (both fade-in and fade-out) would resolve this? Perhaps I'm oversimplifying this, though... or am unaware of how that might affect other things. I also wouldn't know how good that would look/work if someone is **not** using FastTransitions, either. Maybe it'd still look fine? 

--------------------

Not a big deal but in certain cutscenes the red cross remains because the game doesn't consider them to be actual cutscenes, such as: 
- The water draining from the stairs after the RPT apartment boss fight

^ There were a few other instances of this that I didn't think to write down, but it's not really a big deal as they're brief, non-story-critical moments. 

--------------------

Lastly, the game froze two times on me. Both when I pressed a floor level on the elevator button in the Hospital and another when I pressed the floor level on the elevator button in the Hotel. I found it odd that both freezes happened when using the elevator. I have _no clue_ (and doubt) it's related to this fix but keep an eye out if the same happens to you. It's most likely my laptop struggling as it's not well equipped to play this game.

**Comment by Polymega** [comment]
Update: The red cross stays visible on v1.1 and vDC. I also got a game-freeze on my Win7/Nvidia PC when using v1.1 after activating the fix (which I'm sure is due to incompatibility with the fix). 

Perhaps the freezes I've had are relevant to this fix? If this proves to be too much of a headache I say best not to continue pursuing a solution. There's definitely more important/relevant fixes that could be looked in to, instead. And the reduction in size of the cross does help to make it less of a nuisance to keep on-screen.

**Comment by AeroWidescreen** [comment]
Thanks, I'll definitely check for that elevator crash.

> Update: The red cross stays visible on v1.1 and vDC. I also got a game-freeze on my Win7/Nvidia PC when using v1.1 shortly after activating the fix and going into a new room (which I'm sure is due to incompatibility with the fix).

Well, it's only going to work with the v1.0 executable because that's what I made the script for. It's going to change the wrong code for v1.1 and DC, and that's always going to cause a crash.

> Perhaps the freezes I've had are relevant to this fix? If this proves to be too much of a headache I say best not to continue pursuing a solution. There's definitely more important/relevant fixes that could be looked in to, instead. And the reduction in size of the cross does help to make it less of a nuisance to keep on-screen.

It's not a problem, I've just been occupied with some stuff ThirteenAG is working on this week so that's why I'm a little slow on this. I have plenty of ideas I can still try. I'll probably give it another look tonight.

**Comment by elishacloud** [comment]
@AeroWidescreen, if you give me the fix working in v1.0 I can modify it to work on the other versions as well.

**Comment by AeroWidescreen** [comment]
[Download](https://mega.nz/#!8ShXEB5L!t4_LryKXFSFxX3YCUKYnyWmaAo2L6ZJs0UELK-M89tA)

@Polymega 

> For the situations above I'd wonder if disabling the red cross during transitions (both fade-in and fade-out) would resolve this? 

This is what I ended up having to do. There's a brief moment between cutscenes where it goes from 'off' to 'on' again. There was nothing I could do to get around that besides disabling the indicator when transitions are active as well. 

> Lastly, the game froze two times on me. Both when I pressed a floor level on the elevator button in the Hospital and another when I pressed the floor level on the elevator button in the Hotel. I found it odd that both freezes happened when using the elevator..

I wasn't able to reproduce this. I tested the hospital elevators and hotel elevators. No crashing on my end. Try again with my newest revision, but again, only test it with the v1.0 executable.

**Comment by Polymega** [comment]
I tested most of the problem areas I mentioned before and this one worked beautifully, Aero! Many thanks for tweaking this some more. :) And really, having the red cross come back after a transition fade isn't bad-looking at all. And with FastTransitions it's practically unnoticeable. This'll be another great, subtle quality of life adjustment, so many thanks again to you and Elisha for implementation when the time comes!

**Comment by AeroWidescreen** [comment]
@elishacloud

To summarize what I'm doing here: if a transition is not active (value is not greater than 00) or a cutscene is not active (value is 00000000), then the health indicator remains enabled. An exception was made for one cutscene where it gets enabled there too.

The `push eax` and `pop eax` are absolutely necessary or else the game will crash. The fact that ` jmp 00476392` doesn't go to the very end of the function, or that I'm not jumping to a code cave at the very beginning of the function was done intentionally as well. Partly because of crashing and partly because of the noise filter and fog breaking.

```
004762D0 jmp 00000000

00000000 push eax
00000001 cmp byte ptr [01F7D490],00
00000002 jg 00000008 // jumps if transition; disables health indicator
00000003 call 004A02B0 // moves cutscene ID to eax
00000004 cmp eax,24
00000005 je 0000000A // jumps if roof fall cutscene; enables health indicator
00000006 cmp eax,00
00000007 je 0000000A // jumps if not a cutscene; enables health indicator

// disables health indicator
00000008 pop eax
00000009 jmp 00476392

// enables health indicator
0000000A pop eax 
0000000B fld dword ptr [ecx+00000140]
0000000C jmp 004762D6
```

**Comment by elishacloud** [comment]
@AeroWidescreen, thanks for the good write-up.  I am working on this now.

> The `push eax` and `pop eax` are absolutely necessary or else the game will crash.

Since we are cutting in at a "random" place in the ASM function, it is always good practice push all modified registers to the stack and then pop them when done.

**Comment by elishacloud** [comment]
@AeroWidescreen, is it possible for you to give me the following memory address for each of the three exe versions?

> `00000001 cmp byte ptr [01F7D490],00`

The memory is completely blank for thousands of bytes in either directions.  I will need to hard code the memory address for each version.

**Comment by AeroWidescreen** [comment]
Sure, but could you try this first instead?

Replace this code

```
cmp byte ptr [01F7D490],00
jg xxxxxxxx // jumps if transition; disables health indicator
```

with this

```
call 004EECB0 // moves transition value to eax
cmp eax,01
jne xxxxxxxx // jumps if transition is enabled
```

**Comment by elishacloud** [comment]
Oh, it looks like I can get the memory ptr address from here 0x004EECB0.  It seems to me that it would be safer to just use the original ASM:

```asm
cmp byte ptr [01F7D490],00
jg xxxxxxxx // jumps if transition; disables health indicator
```

And then just read the memory address from here 0x004EECB0.

**Comment by AeroWidescreen** [comment]
Awesome. I agree that would be better.

**Comment by Polymega** [comment]
Well done, gentlemen!

**Comment by Polymega** [comment]
## General Health Indicator Improvement:

On each new area/room load, the low health indicator instantly spawns fully opaque and then animates in cycle from opaque to transparent. Because it instantly starts off fully opaque, it looks a little jarring every time it just "pops" up on each new area load. This change will reverse the animation process; to have the low health indicator spawn fully transparent, then animate accordingly, which looks much smoother and intentional:

On game launch:
- `sh2pc.exe+76C7 (004076C7) mov [007F3E14],3F800000` change to `mov [007F3E14],00000000`

This can be generically included in the module; doesn't need its own fix.

## `DisableRedCrossInCutScenes` Improvements

### For "cutscenes"

Some parts of the game that should technically be considered a cutscene (or part of a cutscene) aren't, so the low health indicator still spawns in these parts. The following below will rectify this. This is to be added to the [DisableRedCrossDuringCutscenes.cpp](https://github.com/elishacloud/Silent-Hill-2-Enhancements/blob/master/Patches/DisableRedCrossDuringCutscenes.cpp) patch.

The formula for this is:

- If Room ID == _____ && [InGameCameraPosY](https://github.com/elishacloud/Silent-Hill-2-Enhancements/commit/192bdc4a15ec02c4b5c561a1f6489afc7d247b7c#diff-415aeefc4459fbd45d6dfde2244b220aR41) == _____
   - Disable red cross
   
([You've recently used this code formula](https://github.com/elishacloud/Silent-Hill-2-Enhancements/blob/master/Patches/FlashlightClockPush.cpp#L58) for the new `FixAptClockFlashlight` fix.)
   
So,in addition to disabling the health indicator as it currently does during cutscenes, also disable it when:

- Room: `0x18` && Camera Y: `-112.6049728f` (Failed push clock)
- Room: `0x15` && Camera Y: `-3315.999023f` (Crossing apartments)
- Room: `0x21` && Camera Y: `-3961f` (Apt fight water draining)
- Room: `0x08` && Camera Y: `150f` (Rosewater Park sign)
- Room: `0x40` && Camera Y: `-349.463501f` (Hospital drain fishing)
- Room: `0x53` && Camera Y: `-1135.397583f` (James fridge)
- Room: `0x53` && Camera Y: `-1441.904907f` (James/Maria fridge)
- Room: `0x08` && (Camera Y: `-630.0f` || (Camera Y: `-1904.75f` && [FullscreenImageEvent](https://github.com/elishacloud/Silent-Hill-2-Enhancements/commit/192bdc4a15ec02c4b5c561a1f6489afc7d247b7c#diff-415aeefc4459fbd45d6dfde2244b220aR40) == `0`)) (James monologue at end of hospital)
- Room: `0x87` && Camera Y: `-860.0f` (Cube head puzzle)

### For map transcriptions
The low health indicator stays on screen whenever James transcribes notes onto his own map, which shouldn't happen. There are four different times within the game that he'll transcribe notes. 

To determine when he transcribes, [can your method of detecting when he transcribes to remove the noise filter](https://github.com/elishacloud/Silent-Hill-2-Enhancements/blob/master/Wrappers/d3d8/IDirect3DDevice8.cpp#L1330-L1340) be carried over to this?

Thank you!

**Comment by elishacloud** [comment]
> Health Indicator Improvement

I implemented all the health indicator improvements.  See testing build below.

> To determine when he transcribes, [can your method of detecting when he transcribes to remove the noise filter](https://github.com/elishacloud/Silent-Hill-2-Enhancements/blob/master/Wrappers/d3d8/IDirect3DDevice8.cpp#L1330-L1340) be carried over to this?

I used the `IsInFullscreenImage` variable from this code for detection.  This means the health indicator is disabled anytime you are in a fullscreen image, including when James is transcribing notes.

Testing module: [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/4709841/d3d8.zip)

**Comment by Polymega** [comment]
Thanks Elisha. Working absolutely perfect on all binaries.

I have one adjustment for this fix... something I overlooked:

Change:

- Room == `0x07` (Crossing apartments)

To:

- Room == `0x07` &&  JamesPosZ == `-83662.25f` (Crossing apartments)

... and that's that!

**Comment by elishacloud** [comment]
Ok, updated the code to fix the 'Crossing apartments' section.  You may also want to checkout the inventory screen.

Testing build:  [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/4720297/d3d8.zip)

**Comment by Polymega** [comment]
Final adjustment for "crossing apartments" section works perfect now. Thank you!

> You may also want to checkout the inventory screen.

![tenor](https://user-images.githubusercontent.com/27395260/83584390-041b0f80-a515-11ea-83e7-f3f75a8b3ab7.gif)

**Comment by elishacloud** [comment]
Implemented in: 69b5c196d456cac34685a3ee1005b43ad5aea14a
